core/services/eth/log_broadcaster.go

```diff
@@ -287,21 +296,41 @@ func (b *logBroadcaster) process(subscription eth.Subscription, chRawLogs <-chan
 }
 
 func (b *logBroadcaster) onRawLog(rawLog eth.Log) {
-	// Skip logs that we've already seen
-	if b.cursor.Initialized &&
-		(rawLog.BlockNumber < b.cursor.BlockIndex ||
-			(rawLog.BlockNumber == b.cursor.BlockIndex && uint64(rawLog.Index) <= b.cursor.LogIndex)) {
-		return
-	}
-
 	for listener := range b.listeners[rawLog.Address] {
-		// Make a copy of the log for each listener to avoid data races
-		listener.HandleLog(rawLog.Copy(), nil)
+		// XXX: warning, n+1
```

Comment:
@se3000 I don't there're really anything we can do (anymore) since each log consumer is responsible for checking the status of a log before acting on it. @spooktheducks and I think that is a better approach than automatically marking logs as "consumed" once they've been delivered to the consumer. So to batch these queries we'd have to do some pretty significant reworking. We could maybe pull all the relevant records and pass around a pointer to it on each log, we could also add logs to a cache once they've been processed and batch update them every X seconds or something. But doing that feels out of scope to me at this point.

@spooktheducks any thoughts?

Reply:
Let's make it work, and then optimize it.